File: negenes/R/negenes.R
#' Estimate the number of essential genes in a genome
#'
#' Estimate, via a Gibbs sampler, the posterior distribution of the number of
#' essential genes in a genome with data from a random transposon mutagenesis
#' experiment. (See the technical report cited below.)
#'
#' @param n.sites A vector specifying the number of transposon insertion sites
#' in each gene (alone).  All elements must by strictly positive.
#' @param counts A vector specifying the number of mutants observed for each
#' gene (alone).  Must be the same length as `n.sites`, and all elements
#' must be non-negative integers.
#' @param n.sites2 A vector specfying the number of transposon insertion sites
#' shared by adjacent genes.  The *i*th element is the number of insertion
#' sites shared by genes *i* and *i*+1.  The last element is for
#' sites shared by genes *N* and 1. If NULL, assume all are 0.
#' @param counts2 A vector specfying the number of mutants shared by adjacent
#' gene (analogous to `n.sites2`). The *i*th element is the number of
#' mutants at sites shared by genes *i* and *i*+1.  The last element
#' is for sites shared by genes *N* and 1. If NULL, assume all are 0.
#' @param n.mcmc Number of Gibbs steps to perform.
#' @param skip An integer; only save every `skip` + 1st step.
#' @param burnin Number of initial Gibbs steps to run (output discarded).
#' @param startp Initial proportion of genes for which no mutant was observed
#' that will be assumed essential for the Gibbs sampler.  (Genes for which a
#' mutant was observed are assumed non-essential; other genes are assumed
#' essential independent with this probability.)
#' @param trace If TRUE, print iteration number occassionally.
#' @param calc.prob If TRUE, return the log posterior probability (up to an
#' additive constant) for each saved iteration.
#' @param return.output If TRUE, include detailed Gibbs results in the output.
#'
#' @return A list with components `n.essential` (containing the total
#' number of essential genes at each iteration of the Gibbs sampler)
#' `summary` (a vector containing the estimated mean, SD, 2.5 percentile
#' and 97.5 percentile of the posterior distribution of the number of essential
#' genes.
#'
#' The next component, `geneprob`, is a vector with one element for each
#' gene, containing the estimated posterior probability that each gene is
#' essential.  These are Rao-Blackwellized estimates.
#'
#' If the argument `calc.prob` was true, there will also be a component
#' `logprob` containing the log (base e) of the posterior probability (up
#' to an additive constant) at each Gibbs step.
#'
#' If the argument `return.output` was true, there will also be a matrix
#' with `n.mcmc` / (`skip` + 1) rows (corresponding to the Gibbs
#' steps) and a column for each gene The entries in the matrix are either 0
#' (essential gene) or 1 (non-essential gene) according to the state of that
#' gene at that step in the Gibbs sampler.
#'
#' @author Karl W Broman, \email{broman@@wisc.edu}
#'
#' @importFrom stats quantile sd
#' @export
#' @useDynLib negenes, .registration=TRUE
#'
#' @seealso [negenes::sim.mutants()], [negenes::Mtb80()]
#'
#' @references
#' - Blades, N. J. and Broman, K. W. (2002) Estimating the number of
#'   essential genes in a genome by random transposon mutagenesis.  Technical
#'   Report MS02-20, Department of Biostatistics, Johns Hopkins University,
#'   Baltimore, MD.
#'   <https://www.biostat.wisc.edu/~kbroman/publications/ms0220.pdf>
#'
#' - Lamichhane et al. (2003) A post-genomic method for predicting
#'   essential genes at subsaturation levels of mutagenesis:
#'   application to Mycobacterium, tuberculosis. Proc Natl Acad Sci USA
#'   100:7213-7218
#'   [doi:10.1073/pnas.1231432100](https://doi.org/10.1073/pnas.1231432100)
#'
#' @keywords models
#'
#' @examples
#' data(Mtb80)
#'
#' # simulate 44% of genes to be essential
#' essential <- rep(0,nrow(Mtb80))
#' essential[sample(1:nrow(Mtb80),ceiling(nrow(Mtb80)*0.44))] <- 1
#'
#' # simulate 759 mutants
#' counts <- sim.mutants(Mtb80[,1], essential, Mtb80[,2], 759)
#'
#' # run the Gibbs sampler without returning detailed output
#' \dontrun{output <- negenes(Mtb80[,1], counts[,1], Mtb80[,2], counts[,2])}
#' \dontshow{output <- negenes(Mtb80[,1], counts[,1], Mtb80[,2], counts[,2],
#'                             n.mcmc=100, skip=0, burnin=0)}
#' # run the Gibbs sampler, returning the detailed output
#' \dontrun{output2 <- negenes(Mtb80[,1], counts[,1], Mtb80[,2], counts[,2], return=TRUE)}
#' \dontshow{output2 <- negenes(Mtb80[,1], counts[,1], Mtb80[,2], counts[,2], return=TRUE,
#'                             n.mcmc=100, skip=0, burnin=0)}
#'
negenes <-
function(n.sites, counts, n.sites2=NULL, counts2=NULL,
         n.mcmc=5000, skip=49, burnin=500,
         startp=1, trace=TRUE,
         calc.prob=FALSE, return.output=FALSE)
{
  n.genes <- length(n.sites)

  # check for errors in the input
  if(length(counts) != n.genes)
    stop("n.sites and counts must be the same length")

  if(is.null(n.sites2)) n.sites2 <- rep(0,n.genes)
  if(is.null(counts2)) counts2 <- rep(0,n.genes)
  if(length(n.sites2) != n.genes)
    stop("n.sites2 and n.sites must be the same length")
  if(length(counts2) != n.genes)
    stop("counts2 and n.sites must be the same length")

  if(any(n.sites<0) || any(counts<0) || any(n.sites2<0) || any(counts2<0))
    stop("n.sites, counts, n.sites2, and counts2 must all be >= 0")

  # replace n.mcmc with number of iterations to be saved
  n.mcmc <- ceiling(n.mcmc/(skip+1))

  # update burnin, since using skip
  burnin <- ceiling(burnin/(skip+1))

  if(n.mcmc <= 0 || burnin < 0 || skip < 0)
    stop("n.mcmc, burnin or skip are incorrectly chosen")

  # genes known to be non-essential
  known <- as.numeric(((counts > 0) | (counts2 > 0) |
                       (c(counts2[n.genes],counts2[-n.genes])>0)))
  notknown <- ((1:n.genes)-1)[known==0]
  n.known <- sum(known)
  n.notknown <- n.genes-n.known

  if(startp < 0 || startp > 1)
    stop("startp must be between 0 and 1")

  temp <- as.numeric(return.output)

  n.mutants <- sum(counts) + sum(counts2)

  output <- .C("R_negenes",
               as.integer(n.genes),
               as.integer(n.mutants),
               as.integer(n.sites),
               # n.sites2 made so that first == last and last == first
               as.integer(c(n.sites2[length(n.sites2)],n.sites2,n.sites2[1])),
               as.integer(known),
               as.integer(n.mcmc),
               as.integer(burnin),
               as.integer(skip),
               output = as.integer(rep(0,(n.mcmc-1)*n.genes*temp+n.genes)),
               n.ess = as.integer(rep(0,n.mcmc)),
               geneprob = as.double(rep(0,n.genes)),
               as.integer(rep(0,n.genes+2)),
               as.integer(n.notknown),
               as.integer(notknown),
               as.integer(calc.prob),
               logprob = as.double(rep(0,n.mcmc)),
               as.integer(return.output),
               as.integer(trace),
               as.double(startp),
               PACKAGE="negenes")

  logprob <- output$logprob
  tot.ess <- output$n.ess
  geneprob <- output$geneprob
  if(return.output) {
    output <- matrix(output$output,ncol=n.genes)
    storage.mode(output) <- "integer"
  }
  summ <- c(mean=sum(geneprob),sd=sd(tot.ess),
            quantile(tot.ess,c(0.025,0.975)))

  if(return.output) {
    if(!calc.prob)
      return(list(n.essential=tot.ess, summary=summ,
                  geneprob=geneprob,output=output))
    else
      return(list(n.essential=tot.ess, summary=summ,
                  geneprob=geneprob,
                  logprob=logprob, output=output))
  }
  else {
    if(!calc.prob)
      return(list(n.essential=tot.ess, summary=summ,
                  geneprob=geneprob))
    else
      return(list(n.essential=tot.ess, summary=summ,
                  geneprob=geneprob,logprob=logprob))
  }
}

File: negenes/src/negenes.c
/**********************************************************************
 *
 * negenes.c
 *
 * copyright (c) 2002, Karl W Broman
 * last modified August, 2002
 * first written June, 2002
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License,
 *     version 3, as published by the Free Software Foundation.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but without any warranty; without even the implied warranty of
 *     merchantability or fitness for a particular purpose.  See the GNU
 *     General Public License, version 3, for more details.
 *
 *     A copy of the GNU General Public License, version 3, is available
 *     at https://www.r-project.org/Licenses/GPL-3
 *
 * C functions for the R/negenes package
 *
 * Contains: R_negenes, negenes, reorg_output, int_permute, random_int
 *
 **********************************************************************/

#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/PrtUtil.h>
#include "negenes.h"

/**********************************************************************
 *
 * negenes
 *
 * Gibbs sampler to estimate the posterior distribution of the number of
 * essential genes in a genome with data from a random transposon
 * mutagenesis experiment.
 *
 * INPUT:
 *     n_genes = Number of different genes
 *     n_mutants = Number of observed viable mutants
 *     n_sites = Number of transposon insertion sites in each gene
 *               [vector of length n_genes]
 *     n_sites2 = Number of transposon insertion sites shared by adjacent genes.
 *               [vector of length n_genes+2]
 *               element 0 = element n_genes; element 1 = element n_genes+1
 *     known = Vector of length n_genes; 1 = mutant observed; 0 = not
 *     n_mcmc  = Number of Gibbs iterations to perform (really, to save)
 *     burnin  = Number of initial Gibbs steps, with results discarded
 *               (really, burnin*(skip+1) initial steps are performed)
 *     skip    = Only return every skip+1st step
 *     output  = 1/0 matrix of dimension (n_genes x n_mcmc)
 *     n_ess   = vector of length n_mcmc
 *     geneprob = Rao-Blackwellized estimates of prob. essential for each
 *                gene [length = n_genes]
 *
 *     curstate  = Workspace of integers, of length n_genes+2
 *     n_w = n_genes - sum(known)
 *     w =         Workspace of length n_w, containing indices of
 *                 genes not known to be non-essential
 *     calcprob  = if 1, calculate the log posterior (up to a scalar multiple)
 *     logprob   = vector of length n.mcmc, to contain log posterior
 *     saveoutput = if 1, fill up all of output; if 0, don't.
 *
 *     trace     = if 1, print trace information
 *     startp    Initial proportion of genes for which no mutant was observed
 *               that will be assumed essential for the Gibbs sampler.
 *
 **********************************************************************/

/* Wrapper for R */
void R_negenes(int *n_genes, int *n_mutants, int *n_sites, int *n_sites2,
	       int *known, int *n_mcmc, int *burnin, int *skip, int *output,
	       int *n_ess, double *geneprob, int *curstate, int *n_w,
	       int *w, int *calcprob, double *logprob, int *saveoutput,
	       int *trace, double *startp)
{
  int **Output;

  reorg_output(*n_mcmc, *n_genes, output, &Output);

  GetRNGstate(); /* get random number seed */

  /* shift n_sites2 and curstate */
  n_sites2++;
  curstate++;

  negenes(*n_genes, *n_mutants, n_sites, n_sites2, known, *n_mcmc,
	  *burnin, *skip, Output, n_ess, geneprob, curstate, *n_w,
	  w, *calcprob, logprob, *saveoutput, *trace, *startp);

  PutRNGstate(); /* put random number seed */
}

/* The actual function */
void negenes(int n_genes, int n_mutants, int *n_sites, int *n_sites2,
	     int *known, int n_mcmc, int burnin, int skip, int **Output,
	     int *n_ess, double *geneprob, int *curstate, int n_w,
	     int *w, int calcprob, double *logprob, int saveoutput,
	     int trace, double startp)
{
  int i, j, s, oldstate;
  int cursum, curnum;
  double p;

  /* start point of Gibbs sampler */
  if(startp<1e-10) {
    for(i=0; i<n_genes; i++)
      curstate[i] = 1;  /* state = 1 -> non-essential */
  }
  else if(startp > 1.0 - 1e-10) {
    for(i=0; i<n_genes; i++) {
      if(known[i]==1) curstate[i] = 1;
      else curstate[i] = 0; /* state = 0 -> essential */
    }
  }
  else { /* random start */
    for(i=0; i<n_genes; i++) {
      if(known[i]==1 || unif_rand() > startp)
	curstate[i] = 1;
      else curstate[i] = 0;
    }
  }
  curstate[n_genes] = curstate[0];
  curstate[-1] = curstate[n_genes-1];

  /* get initial counts */
  cursum = curnum = 0;
  for(i=0; i<n_genes; i++) {
    curnum += curstate[i];
    cursum += (curstate[i]*n_sites[i] +
	       curstate[i]*curstate[i+1]*n_sites2[i]);
  }

  /* begin Gibbs iterations */
  for(i=-burnin; i<n_mcmc; i++) {
    for(s=0; s<=skip; s++) {

      /* consider genes in random order */
      int_permute(w, n_w);

      for(j=0; j<n_w; j++) {

	p = gibbsProb(curstate[w[j]], curstate[w[j]-1], curstate[w[j]+1],
		      n_sites[w[j]], n_sites2[w[j]], n_sites2[w[j]-1],
		      curnum, cursum, n_mutants, n_genes);

	oldstate = curstate[w[j]];

	/* simulate new state */
	if(unif_rand() < p)  curstate[w[j]] = 1;
	else curstate[w[j]] = 0;

	/* make curstate wrap around */
	if(w[j]==0) curstate[n_genes] = curstate[0];
	else if(w[j]==n_genes-1) curstate[-1] = curstate[n_genes-1];

	/* update curnum and cursum */
	curnum += (curstate[w[j]]-oldstate);

	cursum += (curstate[w[j]]-oldstate)*
	  (n_sites[w[j]] + curstate[w[j]+1]*n_sites2[w[j]] +
	   curstate[w[j]-1]*n_sites2[w[j]-1]);

      } /* end loop over genes */
    } /* end loop over skips */

    /* if after burnin period, save result */
    if(i >= 0) {
      n_ess[i] = 0;

      for(j=0; j<n_genes; j++) {
	/* Calculate Gibbs prob for Rao-Blackwellized ests */
	if(known[j]) p=1.0; /* mutant observed => non-essential */
	else p = gibbsProb(curstate[j],curstate[j+1],curstate[j-1],
			   n_sites[j], n_sites2[j], n_sites2[j-1],
			   curnum, cursum, n_mutants, n_genes);

	n_ess[i] += curstate[j];
	geneprob[j] += (1-p);
	if(saveoutput) Output[j][i] = curstate[j];
      }
      n_ess[i] = n_genes - n_ess[i]; /* current number essential */

      if(calcprob) {
	logprob[i] = -(double)n_mutants * log((double)cursum);
	logprob[i] += lgammafn((double)curnum+1) + lgammafn((double)(n_genes-curnum+1));
      }
    }

    if(trace && (i+1 == ((i+1)/20)*20))  /* print iteration */
      Rprintf("%4d\n", i+1);

  } /* end loop over gibbs iterations */

  /* update geneprob */
  for(j=0; j<n_genes; j++) geneprob[j] /= (double)n_mcmc;
}



/**********************************************************************
 *
 * gibbsProb: Calculate Gibbs step probability
 *
 * curstate      = current state at current position
 * curstate_next = current state at next position
 * curstate_prev = current state at prev position
 * n_sites       = no. sites for current position
 * n_sites2      = no. sites shared by current and next position
 * n_sites2_prev = no. sites shared by current and prev position
 * curnum        = current number of non-essential genes
 * cursum        = current number of viable sites
 * n_mutants     = number of mutants observed
 * n_genes       = number of genes
 **********************************************************************/
double gibbsProb(int curstate, int curstate_next, int curstate_prev,
		 int n_sites, int n_sites2, int n_sites2_prev,
		 int curnum, int cursum, int n_mutants, int n_genes)
{
  int Ai, Bi;
  double p;

  Ai = curnum - curstate;
  Bi = cursum - curstate*n_sites - curstate*curstate_next*n_sites2 -
    curstate_prev*curstate*n_sites2_prev;

  p = Bi + n_sites + n_sites2*curstate_next + n_sites2_prev*curstate_prev;

  p = exp(-(double)n_mutants * (log((double)p) - log((double)Bi)));

  /* p = Pr(gene is non-essential */
  return(p / (p + (double)(n_genes - Ai) / (double)(Ai + 1)));
}


/* reorganize output vector so it is a doubly indexed array rather
   than a single long vector; afterwards, indexed as
   Output[gene][iteration] */
void reorg_output(int n_mcmc, int n_notobs, int *output, int ***Output)
{
  int i;

  *Output = (int **)R_alloc(n_notobs, sizeof(int *));
  (*Output)[0] = output;
  for(i=1; i<n_notobs; i++)
    (*Output)[i] = (*Output)[i-1] + n_mcmc;

}

/* same as reorg_output, but for doubles
   than a single long vector; afterwards, indexed as
   Output[gene][iteration] */
/***************
** This is no longer needed **
void reorg_output_dbl(int n_mcmc, int n_notobs, double *output, double ***Output)
{
  int i;

  *Output = (double **)R_alloc(n_notobs, sizeof(double *));
  (*Output)[0] = output;
  for(i=1; i<n_notobs; i++)
    (*Output)[i] = (*Output)[i-1] + n_mcmc;

}
****************/

/**********************************************************************
 *
 * int_permute
 *
 *   This function randomly permutes a vector of integers
 *
 * Input:
 *
 *   array = vector of ints; on output, it contains a random
 *           permutation of the input vector
 *
 *   len   = length of the vector
 *
 **********************************************************************/

void int_permute(int *array, int len)
{
  int i, which;
  int tmp;

  for(i=0; i < len; i++) {
    which = random_int(i, len-1);
    tmp = array[which];
    array[which] = array[i];
    array[i] = tmp;
  }
}

/**********************************************************************
 *
 * random_int
 *
 * Generates a random int integer between "low" and "high", inclusive.
 *
 *  Input:
 *
 *    low
 *
 *    high
 *
 **********************************************************************/

int random_int(int low, int high)
{
  return((int)(unif_rand()*(double)(high - low + 1)) + low);
}


/* end of negenes.c */



